### PR TITLE
Code refactoring

### PR DESCRIPTION
**Pull request by agneta1991**
### This is a collaborative project and the participants are Agneta Pupienyte & Joy Taribagshaw

# Within this branch we have:

- We went through the main.rb file and made sure its within the rules of YAGNI, KISS, and DRY
- We have followed SOLID principles, and refactured main.rb file

### CODE REVIEW

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi @agneta1991, @JoyTaribagshaw, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.



## Invalid Code Review Request

We have noticed that the project you have submitted does not have a proper linter setup. Although you have the `robocup` linter file, we have noticed that you have edited some settings, which are not in compliance with the project requirements. As per the [project requirements](https://github.com/microverseinc/linters-config/tree/master/ruby#:~:text=with%20a%20dot.-,Do%20not%20make%20any%20changes%20in%20config%20files%20%2D%20they%20represent%20style,in%20this%20repository%20and%20let%20your%20code%20reviewer%20know%20about%20it.,-When%20you%20open), you should not edit the linters file. We kindly request you to use the proper linter file and update your code accordingly. 

## 🔴🚫Attention🚫🔴 

📢 Your Code Review Request will be marked as invalid🚫 in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadata3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_**📢💡Invalid Code Review Request **does not count** into the code reviews limit.💡**_

File: app.rb [hunk 1]
```diff
@@ -22,33 +22,32 @@ def list_people
     end
   end
 
+  def ask_name_and_age
+    print 'Age:'
+    @age = gets.chomp.to_i
+    print 'Name:'
+    @name = gets.chomp
+  end
```
Comment (mikethreels) [inline app.rb:30]: - [ ] Please have another look at [the Best Practices and Solid lesson](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid), all requests for input and input validation should've been placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input.

File: app.rb [hunk 2]
```diff
@@ -22,33 +22,32 @@ def list_people
...
+
   def create_person
     puts "\nWould you like to create a student(1) or a teacher(2)?"
     option = gets.chomp.to_i
+    ask_name_and_age
```
Comment (mikethreels) [inline app.rb:36]: - [ ] Please have another look at [the Best Practices and Solid lesson](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid), all requests for input and input validation should've been placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input.

File: app.rb [hunk 3]
```diff
@@ -22,33 +22,32 @@ def list_people
...
-      name = gets.chomp
+    if option == 1
 
-      if age < 18
+      if @age < 18
         print 'Has parent permission? [Y/N]:'
         permission = gets.chomp
       end
-      student = Student.new(age, name, parent_permission: permission)
+      student = Student.new(@age, @name, parent_permission: permission)
```
Comment (mikethreels) [inline app.rb:43]: - [ ] Please have another look at [the Best Practices and Solid lesson](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid), all requests for input and input validation should've been placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input.

File: app.rb [hunk 4]
```diff
@@ -22,33 +22,32 @@ def list_people
...
-      print 'Age:'
-      age = gets.chomp.to_i
-      print 'Name:'
-      name = gets.chomp
+    elsif option == 2
       print 'Specialization:'
       specialization = gets.chomp
-      teacher = Teacher.new(specialization, age, name, parent_permission: permission)
+      teacher = Teacher.new(specialization, @age, @name, parent_permission: permission)
       @people.push(teacher)
```
Comment (mikethreels) [inline app.rb:52]: - [ ] Please have another look at [the Best Practices and Solid lesson](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid), all requests for input and input validation should've been placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input.

File: app.rb [hunk 5]
```diff
@@ -67,33 +66,39 @@ def create_book
...
+    if @books.length.positive?
+      puts 'Please select the book from the list below by a number:'
+      @books.each_with_index do |book, index|
+        puts "#{index + 1}. Title: \"#{book.title}\", Author: #{book.author}"
+      end
+      @book_choice = gets.chomp.to_i
+    else
+      puts 'No books added to the list'
+      nil
+    end
```
Comment (mikethreels) [inline app.rb:79]: - [ ] Please have another look at [the Best Practices and Solid lesson](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid), all requests for input and input validation should've been placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input.

File: app.rb [hunk 6]
```diff
@@ -67,33 +66,39 @@ def create_book
...
+  end
+
+  def select_person
     if @people.length.positive?
       puts 'Please select a person from the list below by a number (and not the id):'
       @people.each_with_index do |person, index|
         puts "#{index + 1}. [#{person.class.name}] Name: #{person.name} ID: #{person.id} Age: #{person.age}"
       end
-      person_choice = gets.chomp.to_i
+      @person_choice = gets.chomp.to_i
```
Comment (mikethreels) [inline app.rb:88]: - [ ] Please have another look at [the Best Practices and Solid lesson](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid), all requests for input and input validation should've been placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input.

File: app.rb [hunk 7]
```diff
@@ -67,33 +66,39 @@ def create_book
...
-
+  def create_rental
+    select_person
+    select_book
     print 'Date (YYYY/MM/DD): '
     date = gets.chomp
-    selected_person = @people[person_choice - 1]
-    selected_book = @books[book_choice - 1]
+    selected_person = @people[@person_choice - 1]
+    selected_book = @books[@book_choice - 1]
```
Comment (mikethreels) [inline app.rb:101]: - [ ] Please have another look at [the Best Practices and Solid lesson](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid), all requests for input and input validation should've been placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Good job setting up the project

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app.rb (hunk 1 above)
Comment (agneta1991) [reply app.rb:30]: @mikethreels Hi Mike, we are slihtly confused on your requested changes. as in the requirements it says **"Make sure that the role of main.rb is only to start your code and ask the user for the _initial choice of options_ (or to quit)."**
and every single piece of code you have highlighted, you asked "_all requests for input and input validation should've been placed in the main.rb file"_ could you please explain a little bit more in depth what you would like us to do? do you mean to take everything from app.rb file and join inside the Main class?

Comment (agneta1991) [reply app.rb:30]: requirements

Comment (agneta1991) [reply app.rb:30]: @mikethreels could you please respond as soon as you see this message, as we would like to implement the changes as quick as possible to be able to move on with the next project.

**Review by Nemwel-Boniface — APPROVED** [review]
## Hello @agneta1991 @shahadat3669  and @JoyTaribagshaw   👋,

I am Nemwel your reviewer.


## _This is an additional review as a result of the second opinion request._

I have received your second opinion request with a complaint about the previous reviewer @mikethreels  Below are my findings:

I have inspected the project as at the time the reviewer @mikethreels  made the review and checked the [project requirements](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_refactor.md)  and in my opinion, the project was ready for approval. It is not clear why the reviewer asked the changes.

Because the requested changes are optional, I think your project can be approved.

As this was the only reason that your project was not approved it will be approved as soon as possible.
Please note that since we will approve it manually so this could take a few hours!

In case anything is not 100% clear tag me @Nemwel-Boniface so that I get the notification.

Keep rocking 🚀!

**Pull request merged**